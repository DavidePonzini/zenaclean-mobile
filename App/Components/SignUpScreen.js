import React, { Component } from 'react'
import { StyleSheet, View, Image, Alert } from 'react-native'
import { Button, Text, Input, Item, Container, Content, Icon } from 'native-base'
import Colors from '../Themes/Colors'
import api from '../Services/ApiService'
import ActionButton from 'react-native-action-button'

export default class SignUpScreen extends Component {
  constructor (props) {
    super(props)
    this.state = {
      inputCV: '',
      inputEmail: '',
      inputPassword: '',
      inputRePassword: '',
      validCV: undefined,
      validEmail: undefined,
      validPassword: undefined,
      validRePassword: undefined
    }
  }

  submit = () => {
    let that = this
    if (this.state.validCV && this.state.validEmail && this.state.validPassword && this.state.validRePassword) {
      api.registerUser(this.state.inputEmail, this.state.inputCV, this.state.inputPassword, res => {
        if (res.status === 'ok') {
          Alert.alert('Registrazione effettuata con successo!', '', [
            { text: 'OK', onPress: () => { that.props.navigation.goBack() } }
          ])
        } else {
          Alert.alert(res.error)
        }
      })
    } else {
      Alert.alert('Campi inseriti non validi')
    }
  }

  checkInputCV = (cv) => {
    this.setState({ inputCV: cv })
    let reg = /^[a-zA-Z]{6}[0-9]{2}[a-zA-Z][0-9]{2}[a-zA-Z][0-9]{3}[a-zA-Z]$/
    reg.test(cv) ? this.setState({ validCV: true }) : this.setState({ validCV: false })
  }

  checkInputEmail = (email) => {
    this.setState({ inputEmail: email })
    let reg = /^\w+([.-]?\w+)*@\w+([.-]?\w+)*(\.\w{2,3})+$/
    reg.test(email) ? this.setState({ validEmail: true }) : this.setState({ validEmail: false })
  }

  checkInputPassword = (password) => {
    this.setState({ inputPassword: password })
    password.length >= 8 ? this.setState({ validPassword: true }) : this.setState({ validPassword: false })
  }

  checkInputRePassword = (rePassword) => {
    this.setState({ inputRePassword: rePassword })
    this.state.inputPassword === rePassword
      ? this.setState({ validRePassword: true }) : this.setState({ validRePassword: false })
  }

  render () {
    return (
      <Container>
        <Content>
          <Image style={styles.logo}
            source={require('../Assets/Images/logo.png')}
          />
          <View style={styles.container}>
            <Item rounded style={styles.width_items}
              success={typeof this.state.validCV !== 'undefined' ? this.state.validCV : undefined}
              error={typeof this.state.validCV !== 'undefined' ? !(this.state.validCV) : undefined}>
              <Icon name='ios-person' />
              <Input
                accessibilityLabel='ssn-signup'
                testID={'ssn-signup'}
                placeholder='Codice Fiscale'
                onChangeText={(cv) => this.checkInputCV(cv)} />
            </Item>
            {typeof this.state.validCV !== 'undefined' && !this.state.validCV && <Text style={styles.errorInputMessage}>{'CV non valido'}</Text> }
            <Item rounded style={[styles.width_items, styles.input_pwd]}
              success={typeof this.state.validEmail !== 'undefined' ? this.state.validEmail : undefined}
              error={typeof this.state.validEmail !== 'undefined' ? !(this.state.validEmail) : undefined}>
              <Icon name='ios-at' />
              <Input
                accessibilityLabel='signup-email'
                testID={'signup-email'}
                placeholder='Email'
                onChangeText={(email) => this.checkInputEmail(email)} />
            </Item>
            {typeof this.state.validEmail !== 'undefined' && !this.state.validEmail && <Text style={styles.errorInputMessage}>{'Email non valida'}</Text> }
            <Item rounded style={[styles.width_items, styles.input_pwd]}
              success={typeof this.state.validPassword !== 'undefined' ? this.state.validPassword : undefined}
              error={typeof this.state.validPassword !== 'undefined' ? !(this.state.validPassword) : undefined}>
              <Icon name='ios-lock' />
              <Input placeholder='Password'
                secureTextEntry
                accessibilityLabel='signup-password'
                testID={'signup-password'}
                onChangeText={(password) => this.checkInputPassword(password)} />
            </Item>
            {typeof this.state.validPassword !== 'undefined' && !this.state.validPassword && <Text style={styles.errorInputMessage}>{'La password deve essere almeno di 8 caratteri'}</Text> }
            <Item rounded style={[styles.width_items, styles.input_pwd]}
              success={typeof this.state.validRePassword !== 'undefined' ? this.state.validRePassword : undefined}
              error={typeof this.state.validRePassword !== 'undefined' ? !(this.state.validRePassword) : undefined}>
              <Icon name='ios-lock' />
              <Input placeholder='Conferma Password'
                secureTextEntry
                accessibilityLabel='signup-confirmbutton'
                testID={'signup-confirmbutton'}
                onChangeText={(rePassword) => this.checkInputRePassword(rePassword)} />
            </Item>
            {typeof this.state.validRePassword !== 'undefined' && !this.state.validRePassword && <Text style={styles.errorInputMessage}>{'Le password non coincidono'}</Text> }
            <Button rounded style={styles.button_sign_in}
              accessibilityLabel='signup-button'
              testID={'signup-button'}
              onPress={() => this.submit()}>
              <Text>Registrati</Text>
            </Button>
          </View>
        </Content>
      </Container>
    )
  }
}
const styles = StyleSheet.create({
  container: {
    paddingTop: 50,
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  },
  logo: {
    marginTop: 20,
    width: 150,
    height: 150,
    alignSelf: 'center'
  },
  button_sign_in: {
    marginTop: 40,
    width: '90%',
    alignSelf: 'center',
    justifyContent: 'center',
    backgroundColor: Colors.accent
  },
  width_items: {
    width: '90%'
  },
  input_pwd: {
    marginTop: 20
  },
  text_footer: {
    fontSize: 15,
    color: 'white'
  },
  errorInputMessage: {
    color: 'red',
    fontSize: 12
  }
})
