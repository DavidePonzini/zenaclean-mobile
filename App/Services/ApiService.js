import Secrets from 'react-native-config'

const baseUrl = Secrets.API_URL

const getAddressFromCoords = ({ lat, lng }, cb) => {
  const googleApiUrl = 'https://maps.googleapis.com/maps/api/geocode/json?address='

  fetch(googleApiUrl + lat + ',' + lng + '&key=' + Secrets.GOOGLE_MAPS_API_KEY)
    .then((response) => response.json())
    .then((responseJson) => cb(responseJson))
}

const getMarkers = (cb) => {
  fetch(baseUrl + 'markers.json')
    .then((response) => response.json())
    .then((responseJson) => { cb(responseJson) })
}

const uploadReport = (infoReport) => {
  return fetch(baseUrl + 'markers.json', {
    method: 'POST',
    body: JSON.stringify(infoReport)
  })
}
export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport
}
