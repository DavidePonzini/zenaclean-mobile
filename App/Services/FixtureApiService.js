const SSNlist = ['LNGSVV60T17H509N', 'HTHHLG90R62B112N', 'BRTHZG56P68C982U']
const emailList = ['prova@gmail.com', 'prova@hotmail.it']
export const user1 = {
  email: emailList[0],
  password: '12345678',
  ssn: SSNlist[0],
  id: '5c5afd7c4be7ff1c8d7daacd'
}

const wallets = {}
wallets[user1.id] = 2
export { wallets }
export const fixtureAddress = 'Via del Camoscio'
export const fixtureMarkers = [{ address: fixtureAddress, title: 'Divano abbandonato', latitude: 44.1234, longitude: 8.1234, description: 'Divano abbandonato in via Gramsci', timestamp: '2018-12-09T14:40:35 -02:00', approved_positive: false, approved_negative: false, _id: '5c6580a33a7f647afccd8920', user_id: user1.id, url: 'images/667856780e39f0bbe7407bfdae603f8e', voted_positive: false, voted_negative: false }]
export const fixtureMessageOk = 'Registrazione effettuata con successo!'
export const fixtureMessageNotOk = 'Utente giÃ  esistente'
export const fixtureUserId = 'verylonguserid'
const totVoti = [32, 45, 21]
export let userId = null

const logInUser = (email, password, cb) => {
  return new Promise(resolve => {
    (user1.email === email && user1.password === password) ? resolve({ status: 'ok' }) : resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      userId = fixtureUserId
      cb(null, { id: userId, email, eth_address: null })
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const getAddressFromCoords = ({ lat, lng }, cb) => {
  return new Promise((resolve) => {
    resolve(fixtureAddress)
  }).then((res) => { cb(null, res) })
}

const getMarkers = (neLat, swLat, swLng, neLng, cb) => {
  return new Promise((resolve) => {
    resolve(JSON.stringify(fixtureMarkers))
  }).then((res) => { cb(JSON.parse(res)) })
}

const uploadReport = (infoReport, cb) => {
  return new Promise((resolve) => {
    resolve({ status: 'ok' })
  }).then(() => {
    cb(null)
  })
}

const registerUser = (email, ssn, password, cb) => {
  return new Promise((resolve) => {
    let checked = false
    for (let i = 0; i < SSNlist.length; i++) {
      if (SSNlist[i] === ssn) checked = true
    }
    for (let i = 0; i < emailList.length; i++) {
      if (emailList[i] === email) checked = true
    }
    if (!checked) { resolve(fixtureMessageOk) }
    if (checked) { resolve(fixtureMessageNotOk) }
  }).then((res) => { cb(null, res) })
}

const voteReport = (v, report, cb) => {
  return new Promise((resolve) => {
    let old = totVoti[report]
    let current = totVoti[report] + 1
    if (old < current) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res, current) => {
    if (res.status === 'ok') {
      totVoti[report] = current
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const changePassword = (email, oldPassword, newPassword, confirmPassword, cb) => {
  return new Promise((resolve) => {
    if (user1.password === oldPassword && newPassword === confirmPassword) resolve({ status: 'ok' })
    else resolve({ status: 'no' })
  }).then((res) => {
    if (res.status === 'ok') {
      user1.password = newPassword
      cb(null)
    } else {
      cb(new Error('login failed'), res)
    }
  })
}

const isLoggedIn = () => userId != null

const logoutUser = () => {
  userId = null
}
const rehydrateLogin = () => {
  return new Promise((resolve, reject) => {
    if (userId != null) {
      resolve(userId)
    } else {
      reject(new Error('Not logged in'))
    }
  })
}

const getBalance = (eth_addr, cb) => {
  return new Promise(resolve => {
    resolve({ value: wallets[user1.id] })
  }).then(res => {
    cb(res.value)
  })
}

export default {
  getAddressFromCoords,
  getMarkers,
  uploadReport,
  registerUser,
  logInUser,
  changePassword,
  voteReport,
  isLoggedIn,
  logoutUser,
  rehydrateLogin,
  getBalance
}
