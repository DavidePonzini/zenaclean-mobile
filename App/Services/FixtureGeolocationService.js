const fixturePosition = {
  coords: {
    accuracy: 1135,
    altitude: null,
    altitudeAccuracy: null,
    heading: null,
    latitude: 44.4135423,
    longitude: 8.9554223,
    speed: null
  },
  timestamp: 1547641567823
}

const fixtureRegion = {
  latitude: 44.4135423,
  longitude: 8.9554223,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

let lastLocation = fixturePosition
let currentRegion = fixtureRegion

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    let pos = fixturePosition
    lastLocation = pos
    resolve(pos)
  })
}

const getLatestPosition = () => lastLocation

const geolocateContinuously = (cb, cbErr, options) => {
  return setInterval(cb, 100)
}

const stopGeolocating = (id) => {
  clearInterval(id)
}

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
