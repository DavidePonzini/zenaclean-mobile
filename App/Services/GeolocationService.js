let lastLocation = null

// initial region
let currentRegion = {
  latitude: 44.4056,
  longitude: 8.9463,
  latitudeDelta: 0.1,
  longitudeDelta: 0.1
}

const setCurrentRegion = reg => { currentRegion = reg }
const getCurrentRegion = () => currentRegion

const geolocateOnce = async () => {
  return new Promise(resolve => {
    navigator.geolocation.getCurrentPosition((pos) => {
      lastLocation = pos
      resolve(pos)
    })
  })
}

const geolocateContinuously = (cb, cbErr, options) => {
  return navigator.geolocation.watchPosition((pos) => {
    cb(pos)
    lastLocation = pos
  }, cbErr, options)
}

const stopGeolocating = (id) => {
  navigator.geolocation.clearWatch(id)
}

const getLatestPosition = () => lastLocation

export default {
  geolocateOnce,
  geolocateContinuously,
  stopGeolocating,
  getLatestPosition,
  getCurrentRegion,
  setCurrentRegion
}
