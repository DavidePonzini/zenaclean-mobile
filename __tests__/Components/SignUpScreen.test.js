import React from 'react'
import SignUpScreen from '../../App/Components/SignUpScreen'
import Alert from 'Alert'

import { user1 } from '../../App/Services/FixtureApiService'
import { shallow, configure } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import 'isomorphic-fetch'
import { shallowToJson } from 'enzyme-to-json'

configure({ adapter: new Adapter() })

jest.mock('../../App/Services/ApiService', () => require('../../App/Services/FixtureApiService'))
jest.mock('Alert', () => {
  return {
    alert: jest.fn()
  }
})
const navigation = { navigate: jest.fn(), goBack: jest.fn() }

describe('SignUpScreen tests', () => {
  const wrapper = shallow(<SignUpScreen navigation={navigation} />)
  it('renders correctly', () => {
    wrapper.update()
    expect(shallowToJson(wrapper)).toMatchSnapshot()
  })

  it('shows a popup on failed sign up attempt', async (done) => {
    Alert.alert.mockClear()
    const ssnInput = wrapper.find({ testID: 'ssn-signup' }).first()
    ssnInput.simulate('changeText', user1[2])
    const emailInput = wrapper.find({ testID: 'signup-email' }).first()
    emailInput.simulate('changeText', user1[0])
    const pwdInput = wrapper.find({ testID: 'signup-password' }).first()
    pwdInput.simulate('changeText', user1[1])
    const cpwdInput = wrapper.find({ testID: 'signup-confirmbutton' }).first()
    cpwdInput.simulate('changeText', 'nop')
    wrapper.update()
    const signUpButton = wrapper.find({ testID: 'signup-button' }).first()
    await signUpButton.simulate('Press')
    await expect(Alert.alert.mock.calls[0][0]).toEqual('Campi inseriti non validi o mancanti')
    done()
  })

  it('navigates when successful register', async (done) => {
    Alert.alert.mockClear()
    const ssnInput = wrapper.find({ testID: 'ssn-signup' }).first()
    ssnInput.simulate('changeText', user1[2])
    const emailInput = wrapper.find({ testID: 'signup-email' }).first()
    emailInput.simulate('changeText', user1[0])
    const pwdInput = wrapper.find({ testID: 'signup-password' }).first()
    pwdInput.simulate('changeText', user1[1])
    const cpwdInput = wrapper.find({ testID: 'signup-confirmbutton' }).first()
    cpwdInput.simulate('changeText', user1[1])
    const gdprButton = wrapper.find({ testID: 'gdpr-checkbox' }).first()
    await gdprButton.simulate('Press')
    const signUpButton = wrapper.find({ testID: 'signup-button' }).first()
    await signUpButton.simulate('Press')
    Alert.alert.mock.calls[0][2][0].onPress()
    await expect(navigation.goBack).toHaveBeenCalled()
    done()
  })
})
