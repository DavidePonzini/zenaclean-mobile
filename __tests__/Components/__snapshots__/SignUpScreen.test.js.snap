// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpScreen tests renders correctly 1`] = `
<Styled(Container)>
  <Styled(Content)>
    <Component
      source={
        Object {
          "testUri": "../../../App/Assets/Images/logo.png",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "height": 150,
          "marginTop": 20,
          "width": 150,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "flex": 1,
          "justifyContent": "center",
          "paddingTop": 50,
        }
      }
    >
      <Styled(Item)
        rounded={true}
        style={
          Object {
            "width": "90%",
          }
        }
      >
        <Styled(Icon)
          name="ios-person"
        />
        <Styled(Input)
          accessibilityLabel="ssn-signup"
          onChangeText={[Function]}
          placeholder="Codice Fiscale"
          testID="ssn-signup"
        />
      </Styled(Item)>
      <Styled(Item)
        rounded={true}
        style={
          Array [
            Object {
              "width": "90%",
            },
            Object {
              "marginTop": 20,
            },
          ]
        }
      >
        <Styled(Icon)
          name="ios-at"
        />
        <Styled(Input)
          accessibilityLabel="signup-email"
          onChangeText={[Function]}
          placeholder="Email"
          testID="signup-email"
        />
      </Styled(Item)>
      <Styled(Item)
        rounded={true}
        style={
          Array [
            Object {
              "width": "90%",
            },
            Object {
              "marginTop": 20,
            },
          ]
        }
      >
        <Styled(Icon)
          name="ios-lock"
        />
        <Styled(Input)
          accessibilityLabel="signup-password"
          onChangeText={[Function]}
          placeholder="Password"
          secureTextEntry={true}
          testID="signup-password"
        />
      </Styled(Item)>
      <Styled(Item)
        rounded={true}
        style={
          Array [
            Object {
              "width": "90%",
            },
            Object {
              "marginTop": 20,
            },
          ]
        }
      >
        <Styled(Icon)
          name="ios-lock"
        />
        <Styled(Input)
          accessibilityLabel="signup-confirmbutton"
          onChangeText={[Function]}
          placeholder="Conferma Password"
          secureTextEntry={true}
          testID="signup-confirmbutton"
        />
      </Styled(Item)>
      <CheckBox
        accessibilityLabel="gdpr-checkbox"
        center={false}
        checked={false}
        checkedColor="green"
        checkedIcon="check-square-o"
        iconRight={false}
        onPress={[Function]}
        right={false}
        size={24}
        testID="gdpr-checkbox"
        title="Acconsento al trattamento dei dati personali"
        uncheckedColor="#bfbfbf"
        uncheckedIcon="square-o"
      />
      <Styled(Text)
        onPress={[Function]}
        style={
          Array [
            Object {
              "color": "dodgerblue",
              "textDecorationLine": "underline",
            },
          ]
        }
      >
         Maggiori informazioni
      </Styled(Text)>
      <Styled(Button)
        accessibilityLabel="signup-button"
        onPress={[Function]}
        rounded={true}
        style={
          Object {
            "alignSelf": "center",
            "backgroundColor": "dodgerblue",
            "justifyContent": "center",
            "marginTop": 40,
            "width": "90%",
          }
        }
        testID="signup-button"
      >
        <Styled(Text)>
          Registrati
        </Styled(Text)>
      </Styled(Button)>
    </View>
  </Styled(Content)>
</Styled(Container)>
`;
